modules/fedimint-ln-client/src/lib.rs

```diff
@@ -25,65 +57,405 @@ impl ClientModuleGen for LightningClientGen {
+            Some(pub_key) => gateways
+                .into_iter()
+                .find(|g| g.node_pub_key == pub_key)
+                .ok_or_else(|| {
+                    debug!("Could not find gateway with public key {:?}", pub_key);
+                    anyhow::anyhow!("Could not find gateway with public key {:?}", pub_key)
+                })?,
+            // Otherwise (no pubkey provided), select and activate the first registered gateway.
+            None => {
+                debug!("No public key for gateway supplied, using first registered one");
+                gateways[0].clone()
+            }
+        };
+        dbtx.insert_entry(&LightningGatewayKey, &gateway).await;
+        Ok(gateway)
+    }
+
+    pub async fn fetch_registered_gateways(
+        &self,
+        global_context: DynGlobalClientContext,
+    ) -> anyhow::Result<Vec<LightningGateway>> {
+        Ok(global_context.api().fetch_gateways().await?)
     }
 }
 
 #[derive(Debug)]
-pub struct LightningClientModule {}
+pub struct LightningClientModule {
+    module_instance_id: ModuleInstanceId,
+    cfg: LightningClientConfig,
+    notifier: ModuleNotifier<DynGlobalClientContext, LightningClientStateMachines>,
+}
 
 impl ClientModule for LightningClientModule {
     type Common = LightningModuleTypes;
-    type ModuleStateMachineContext = ();
-    type States = LightningClientStates;
+    type ModuleStateMachineContext = LightningClientContext;
+    type States = LightningClientStateMachines;
 
-    fn context(&self) -> Self::ModuleStateMachineContext {}
+    fn context(&self) -> Self::ModuleStateMachineContext {
+        LightningClientContext {
+            module_instance_id: self.module_instance_id,
+            secp: secp256k1_zkp::Secp256k1::new(),
+        }
+    }
 
-    fn input_amount(
-        &self,
-        _input: &<Self::Common as ModuleCommon>::Input,
-    ) -> TransactionItemAmount {
-        unimplemented!()
+    fn input_amount(&self, input: &<Self::Common as ModuleCommon>::Input) -> TransactionItemAmount {
+        TransactionItemAmount {
+            amount: input.amount,
+            fee: self.cfg.fee_consensus.contract_input,
+        }
     }
 
     fn output_amount(
         &self,
-        _output: &<Self::Common as ModuleCommon>::Output,
+        output: &<Self::Common as ModuleCommon>::Output,
     ) -> TransactionItemAmount {
-        unimplemented!()
+        match output {
+            LightningOutput::Contract(account_output) => TransactionItemAmount {
+                amount: account_output.amount,
+                fee: self.cfg.fee_consensus.contract_output,
+            },
+            LightningOutput::Offer(_) | LightningOutput::CancelOutgoing { .. } => {
+                TransactionItemAmount {
+                    amount: Amount::ZERO,
+                    fee: Amount::ZERO,
+                }
+            }
+        }
+    }
+}
+
+impl LightningClientModule {
+    pub async fn fetch_active_gateway(
```

Comment:
Not thrilled that I needed to replicate these functions to expose them to `LightningClientModule`. Perhaps there's a better spot to put these

Reply:
Refactored out of `LightningClientContext`